core/src/main/java/org/apache/iceberg/FileCleanupStrategy.java

```diff
@@ -79,4 +80,15 @@ protected void deleteFiles(Set<String> pathsToDelete, String fileType) {
             (file, thrown) -> LOG.warn("Delete failed for {} file: {}", fileType, file, thrown))
         .run(deleteFunc::accept);
   }
+
+  protected Set<String> expiredStatisticsFilesLocations(
+      TableMetadata beforeExpiration, Set<Long> expiredIds) {
+    Set<String> expiredStatisticsFilesLocations = Sets.newHashSet();
+    for (StatisticsFile statisticsFile : beforeExpiration.statisticsFiles()) {
+      if (expiredIds.contains(statisticsFile.snapshotId())) {
+        expiredStatisticsFilesLocations.add(statisticsFile.path());
```

Comment:
- `statisticsFile.snapshotId()` is "ID of the Iceberg table's snapshot the statistics were computed from",  it's not "ID of the snapshot the file is attached to"
- The spec allows one Puffin file to be attached to multiple snapshots. For example, rewrite_files / optimize will produce a new snapshot, but could carry forward some/all table-level stats.
  - So, the fact that the snapshot is being forgotten doesn't imply the stats file can be safely deleted.

Reply:
> The spec allows one Puffin file to be attached to multiple snapshots. For example, rewrite_files / optimize will produce a new snapshot, but could carry forward some/all table-level stats.
So, the fact that the snapshot is being forgotten doesn't imply the stats file can be safely deleted.

makes sense. I will compute live statistics file set and then check whether it can be safely deleted along with the current changes.

Reply:
@findepi: Thinking more about this, As the `TableMetadata` has just the list of  `StatisticsFile`. And you have mentioned,  statisticsFile.snapshotId() is "ID of the Iceberg table's snapshot the statistics were computed from"

So, how will the query knows which statistics file to use for the current snapshot (Incase of rewrite data files, the current snapshot id may not be present in that list of statistics file?)